src/Testing/CurlRunnings/Internal/KeyValuePairs.hs

```diff
@@ -0,0 +1,46 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | A module defining the KeyValuePairs type. This type may be used to
+-- represent a structure in a specification that is a collection of
+-- key-vaue pairs. For example query parameters and URLEncoded request
+-- bodies.
+--
+-- The module provides FromJSON and ToJSON instances for KeyValuePairs.
+-- Valid KeyValuePairs JSON is a single JSON object where all values
+-- are either String, Scienfific or Bool.
+
+module Testing.CurlRunnings.Internal.KeyValuePairs
+  ( KeyValuePairs (..)
+  , KeyValuePair (..)
+  ) where
+
+import           Data.Aeson
+import           Data.Aeson.Types
+import qualified Data.ByteString.Lazy as LBS
+import           Data.HashMap.Strict  as H
+import qualified Data.Text            as T
+import           Data.Text.Encoding   as T
+
+-- | A container for a list of key-value pairs
+newtype KeyValuePairs = KeyValuePairs [KeyValuePair] deriving Show
+
+-- | A representation of a single key-value pair
+data KeyValuePair = KeyValuePair T.Text T.Text deriving Show
+
+instance ToJSON KeyValuePairs where
+  toJSON (KeyValuePairs qs) =
+    object (fmap (\(KeyValuePair k v) -> k .= toJSON v) qs)
+
+instance FromJSON KeyValuePairs where
+  parseJSON = withObject "keyValuePairs" parseKeyValuePairs where
+    parseKeyValuePairs o = KeyValuePairs <$> traverse parseKeyValuePair (H.toList o)
+    parseKeyValuePair (t, v) = KeyValuePair t <$> parseSingleValueType v
+
+parseSingleValueType :: Value -> Parser T.Text
+parseSingleValueType (Bool b)   = parseToText b
+parseSingleValueType (String t) = return t
+parseSingleValueType (Number n) = parseToText n
+parseSingleValueType invalid    = typeMismatch "KeyValuePairs" invalid
```

Comment:
Was the omission of arrays in the interest of just keeping it simple? I can imagine that's a bit more complicated to implement so definitely not necessary, just wondering.

Reply:
Thanks for the review.

My intention for the `KeyValuePairs` type is to model things like query parameters, URLEncoded bodies and Headers. Each case is a collection of key-value pairs where the value is not a container.  So in my model I want to exclude container values (object, array). For example values in a URLEncoded body cannot be arrays or JSON objects - so the model excludes this case.

Maybe the name could be improved? `SingleValuedKeyValuePairs` ?

Reply:
Got it. So I have seen URLEncoded bodies specify array-like values like so: `a_list[]=first_val&a_list[]=second_val&...`. But when I looked into it I see that's a thing some implementations have but it's not in the standard. TIL.

Fortunately, this can still encode a list like that by specifying key-val pairs with the same key. So this should be all good either way.

Reply:
And that's ok about the name change, I think that name is good enough as-is.